### ISSUE TITLE
conflict between prefer-const and no-use-before-define (2.5.0 regression)

### ISSUE THREAD

**Issue by lukeapage** [eslint/eslint#5692]
**What version of ESLint are you using?**
2.5.0 (I think my previous version that didn't fail was 2.4.0, but it could have been a different 2.x)
**What parser (default, Babel-ESLint, etc.) are you using?**
Default
**Please show your full configuration:**
```js
{
...
"no-use-before-define": ["error", "nofunc"],
...
 "prefer-const": "error",
}
```
**What did you do? Please include the actual source code causing the issue.**
```js
            let interval;
            function checkForWindow() {
                const newFocussed = BrowserWindow.getFocusedWindow();
                if (newFocussed !== currentlyFocussed) {
                    clearInterval(interval);
                    newFocussed.toggleDevTools();
                }
            }
            interval = setInterval(checkForWindow, msIntervalToCheckForWindow);
```
**What did you expect to happen?**
no warnings
**What actually happened? Please include the actual, raw output from ESLint.**
either
```sh
  76:13  error  'interval' is never reassigned, use 'const' instead  prefer-const
```
or if I change it to const after the interval
```sh
  72:35  error  'interval' was used before it was defined    no-use-before-define
```

I see the rule says this is an error
```js
// the initializer is separated.
let a;
a = 0;
console.log(a);
```
but this is not
```js
let a;
if (true) {
    a = 0;
}
console.log(a);
```
I would suggest that
```js
let a;
// read or write of a in any sub block (e.g. if or function)
console.log(a);
```
is not an error?

**Comment by lukeapage** [comment]
work-around is to un-necessarily assign null to the let declaration.

**Comment by alberto** [comment]
ping @mysticatea

**Comment by mysticatea** [comment]
Thank you for this issue.

I'm sorry for late.

Surely, the 2 rules can conflict.
But in this case, actually, these are not conflicting because you are using `"nofunc"` option of `no-use-before-define` rule.

```js
/*eslint
  	no-use-before-define: ["error", "nofunc"],
  	prefer-const: "error"
 */

const interval = setInterval(checkForWindow, msIntervalToCheckForWindow);
function checkForWindow() {
    const newFocussed = BrowserWindow.getFocusedWindow();
    if (newFocussed !== currentlyFocussed) {
        clearInterval(interval);
        newFocussed.toggleDevTools();
    }
}
```

I think we can do 2 things.

- Moving the location of the error to the declaration if the reference was read before the assignment.
- Adding an option of `prefer-const` rule to allow `let` declaration if the reference was read before the assignment.

**Comment by lukeapage** [comment]
for the 2nd option, its only valid (in the case of `let`) inside a function because of TDZ so it seems a very specific extra option to work-around the rule being overly aggressive - it doesn't seem to me to be that much of a relaxation to not create an error if its read in a sub function scope.
I'm in 2 minds as to what is most important - define before use or mark variable as const - because thats what the end user of eslint has to decide in these situations when using the 2 rules together and encountering the above.
Probably my work-around to assign null is the best thing to do if you don't want to change beaviour.

**Comment by bighuggies** [comment]
However this was working previously was perfect. What errors are now being caught that weren't before because of this?

**Comment by mysticatea** [comment]
@bighuggies Before, when the initialization was separated, `prefer-const` had been warning variables even if the assignment exists in another block. So it had a different logic with now. The cases that the assignment exists in another block will be separated to another rule: #5284 

----

I try to more explain my 2 proposal. I intended both 2 proposals apply.

> - Moving the location of the error to the declaration if the reference was read before the assignment.

Currently, `prefer-const` shows errors at the initialization/assignment of variables because I guess we write `const` there to fix.
But if reading references exist before the assignment, we need to move the assignment to the declaration then modify `let` with `const`. This is the sample code of this issue.
So I proposed moving the location of the error to the declaration if the reading reference exist before the assignment.

> - Adding an option of prefer-const rule to allow let declaration if the reference was read before the assignment.

If we use `no-use-before-define` without `"nofunc"` option, the current behavior of `prefer-const` can conflict with that. @lukeapage's workaround would work fine, but I think good as `prefer-const` has the option to avoid the conflict.

**Comment by ljharb** [comment]
@mysticatea closed #5822 as a dupe - any update on this fix?

**Comment by mysticatea** [comment]
I'm sorry for late response.
I will do within next week. I have a vacation in next week!!

**Comment by nzakas** [comment]
I'm wondering -- should this be the default behavior rather than adding an option?

**Comment by ljharb** [comment]
Yes, I think it should, but would that make it a breaking change?

**Comment by nzakas** [comment]
See note on the PR. Option makes sense.
